### PR TITLE
[GDN] Add Hopper SM90 CuTe DSL prefill

### LINKED ISSUE #76: [Feature] Integrate Lightning SM90 Implementation via CuTe DSL
# [Feature] Integrate Lightning SM90 Implementation via CuTe DSL

## Background

FlashInfer has a production-ready Lightning Attention SM90 (Hopper) implementation. For API completeness, we could port it into cuLA using **CuTe DSL**, along with corresponding benchmarks and unit tests.

Upstream PR: https://github.com/flashinfer-ai/flashinfer/pull/2276 from guangyunh-nv

> **Note**: The kernel design and algorithm credits belong to the original authors (guangyunh-nv). This issue tracks a porting effort, not an original implementation.

## Upstream Implementation Overview

FlashInfer PR #2276 implements Lightning Attention Prefill on Hopper architecture:

- **SM90 Optimized**: TMA warp-specialized architecture with asynchronous copy and warp group scheduling
- **Optional gating** and final-state output
- **High-level Python API** with chunked prefill support

## Task Checklist

- [ ] **Port kernels via CuTe DSL**: Rewrite the upstream C++ Hopper lightning kernels using CuTe DSL, following cuLA's build system and coding conventions
- [ ] **Adapt Python interface**: Align with cuLA's existing API style and expose the Lightning SM90 Python bindings
- [ ] **Add benchmarks**: Add performance benchmarks under cuLA's framework, covering the same settings as the KDA benchmarks:
  - Fixed-length: `B={1,2}, T={512, 1024, 4096, 8192, 16384}`, `H=64, D=128, dtype=bf16`
  - Varlen: `num_seqs={10, 20}, total_len={4096, 8192, 16384}`, distributions: uniform / random / skewed
- [ ] **Performance validation**: Verify that the CuTe DSL version achieves comparable performance to the upstream C++ version across all benchmark settings above
- [ ] **Add unit tests**: Port the reference implementation and test cases from upstream `tests/gdn/` to ensure correctness
- [ ] **Update documentation**: Document the newly added Lightning SM90 support

## References

- Upstream PR: [flashinfer-ai/flashinfer#2276](https://github.com/flashinfer-ai/flashinfer/pull/2276)
- Related Issue: [flashinfer-ai/flashinfer#1690](https://github.com/flashinfer-ai/flashinfer/issues/1690)

### PR DESCRIPTION
**Pull request by Aharrypotter**
## 📌 Description

This PR adds a Hopper SM90 Gated DeltaNet prefill path implemented entirely in
CuTe DSL.

- ports the upstream 512-thread TMA/WGMMA warp-specialized algorithm into
  `cula.ops.gdn.sm90` while preserving upstream authorship and provenance;
- exposes packed variable-length BF16 prefill through
  `cula.gdn.chunk_gated_delta_rule`, including optional gates, initial/final
  state, MHA, GQA, and GVA head mappings;
- keeps the product path independent of the cuLA C++ extension and FlashInfer
  headers/submodules at runtime;
- adds the exact 28-row Issue #76 benchmark, independent PyTorch reference
  tests, safety coverage, and Hopper user documentation.

## 🔍 Related Issues

Closes #76.

## Issue checklist

| Issue requirement | Delivery |
| --- | --- |
| Port Hopper kernels through CuTe DSL | `cula/ops/gdn/sm90/` implements the SM90a 512-thread TMA/WGMMA kernel and direct TVM-FFI launch path. |
| Adapt the Python interface | `cula.gdn.chunk_gated_delta_rule` validates packed inputs and exposes optional gates and recurrent state. |
| Add benchmarks | `benchmarks/bench_gdn_prefill.py` defaults to all 10 fixed and 18 packed-varlen rows from the issue. |
| Validate performance | H20 three-backend, three-process comparison: `0.993329822x` geometric mean and `1.147581166x` worst row versus frozen upstream C++; all 28 rows pass the `1.10x` / `1.25x` gates. |
| Add unit tests | Independent blockwise reference plus C01–C14, head mappings, tails, redzones, repeat, continuation, and stream behavior. |
| Update documentation | README quick start, repository layout, API/requirements guide, and pipeline guide. |

### ✅ Pre-commit Checks

- [ ] I have run the repository-wide pre-commit hook set.
- [x] Ruff lint and format checks pass for every changed Python file.
- [x] `git diff --check` passes.

## 🧪 Tests

- [x] H20/SM90 product suite: `22 passed`.
- [x] C01–C14 correctness, no-C++ dispatch, non-default stream,
  repeat/continuation, and output/state redzones pass.
- [x] Compute Sanitizer C04 ragged-tail gate: memcheck and synccheck report
  zero errors; racecheck reports zero hazards.
- [x] Source manifest matches before and after validation (`182/182` files).
- [x] Fresh SM90a codegen: 512 threads, 100 WGMMA/HGMMA, 22 TMA,
  128 registers, zero stack/spill/local memory, and one active CTA/SM.

## ⚡ Performance

Measured on one idle NVIDIA H20-3e with BF16 `Hq=Hk=Hv=64`, `D=128`, 20
warmups and 100 CUDA-event timed iterations per row. Each backend ran in three
independent processes; each table entry is the median of the three process
averages. Compilation and first-call setup are excluded. A ratio below `1.0x`
means the CuTe DSL kernel is faster.

| Check | Limit | Result |
| --- | ---: | ---: |
| DSL / frozen FlashInfer geometric mean | `<= 1.10x` | `0.993329822x` |
| Worst row | `<= 1.25x` | `1.147581166x` (`fixed-b1-t512`) |
| Coverage | 28 rows/backend | DSL, migrated CUTLASS, and frozen FlashInfer each `28/28` |

| Workload group | Rows | DSL / FlashInfer geomean | DSL / migrated CUTLASS geomean | Max DSL / FlashInfer |
| --- | ---: | ---: | ---: | ---: |
| Fixed length | 10 | `1.003484x` | `0.972185x` | `1.147581x` |
| Packed variable length | 18 | `0.987733x` | `0.976909x` | `0.999400x` |
| **Overall** | **28** | **`0.993330x`** | **`0.975219x`** | **`1.147581x`** |

<details>
<summary>Full 28-row median latency table</summary>

| Row | CuTe DSL ms | Migrated CUTLASS ms | Frozen FlashInfer ms | DSL / FlashInfer |
| --- | ---: | ---: | ---: | ---: |
| `fixed-b1-t512` | 0.058701 | 0.056546 | 0.051152 | 1.147581x |
| `fixed-b1-t1024` | 0.097943 | 0.106133 | 0.099455 | 0.984794x |
| `fixed-b1-t4096` | 0.373638 | 0.383458 | 0.377599 | 0.989509x |
| `fixed-b1-t8192` | 0.741560 | 0.754011 | 0.748508 | 0.990718x |
| `fixed-b1-t16384` | 1.477579 | 1.499711 | 1.494760 | 0.988506x |
| `fixed-b2-t512` | 0.100763 | 0.110224 | 0.102313 | 0.984847x |
| `fixed-b2-t1024` | 0.192742 | 0.202722 | 0.195238 | 0.987216x |
| `fixed-b2-t4096` | 0.743809 | 0.757794 | 0.751203 | 0.990157x |
| `fixed-b2-t8192` | 1.480442 | 1.497819 | 1.492454 | 0.991951x |
| `fixed-b2-t16384` | 2.951255 | 2.984504 | 2.980956 | 0.990037x |
| `varlen-n10-t4096-uniform` | 0.385270 | 0.400514 | 0.392410 | 0.981806x |
| `varlen-n10-t4096-random` | 0.354295 | 0.365720 | 0.358675 | 0.987790x |
| `varlen-n10-t4096-skewed` | 0.351345 | 0.361929 | 0.355406 | 0.988573x |
| `varlen-n10-t8192-uniform` | 0.696737 | 0.713239 | 0.705871 | 0.987060x |
| `varlen-n10-t8192-random` | 0.668328 | 0.683563 | 0.675664 | 0.989143x |
| `varlen-n10-t8192-skewed` | 0.653087 | 0.665490 | 0.659418 | 0.990399x |
| `varlen-n10-t16384-uniform` | 1.366857 | 1.383636 | 1.378220 | 0.991755x |
| `varlen-n10-t16384-random` | 1.273788 | 1.291183 | 1.285643 | 0.990779x |
| `varlen-n10-t16384-skewed` | 1.247147 | 1.264023 | 1.258424 | 0.991039x |
| `varlen-n20-t4096-uniform` | 0.430677 | 0.447858 | 0.439974 | 0.978869x |
| `varlen-n20-t4096-random` | 0.408003 | 0.420942 | 0.413756 | 0.986093x |
| `varlen-n20-t4096-skewed` | 0.401375 | 0.413276 | 0.406397 | 0.987642x |
| `varlen-n20-t8192-uniform` | 0.725918 | 0.747053 | 0.739044 | 0.982240x |
| `varlen-n20-t8192-random` | 0.701505 | 0.721255 | 0.713559 | 0.983107x |
| `varlen-n20-t8192-skewed` | 0.707668 | 0.714869 | 0.708092 | 0.999400x |
| `varlen-n20-t16384-uniform` | 1.315303 | 1.338503 | 1.331811 | 0.987604x |
| `varlen-n20-t16384-random` | 1.345815 | 1.369917 | 1.363760 | 0.986842x |
| `varlen-n20-t16384-skewed` | 1.298094 | 1.317645 | 1.312225 | 0.989232x |

</details>

All backend/process inputs are byte-identical. The canonical benchmark can be
reproduced with:

```bash
python benchmarks/bench_gdn_prefill.py --output gdn-sm90-issue76.json
```

## Reviewer Notes

- The kernel is adapted from FlashInfer's Apache-2.0 SM90 CuTe DSL rewrite:
  original rewrite `eb31811856e07f3c5fc94a7aaafeaeaabd954f4f`, imported source
  `97996a4123924140d28a3079292535aa534eba92`, and frozen C++ oracle
  `41e5aa29a0e218bde2e2045316ed88e3f4dc8ca2`.
- The imported revision includes the corrected final-tail `s < B and t < B`
  predicate and explicit compile-cache key; cuLA supplies the public API,
  packaging, tests, benchmark, and documentation around the upstream kernel.
- The SM90 path requires `nvidia-cutlass-dsl==4.5.1` and fails explicitly on an
  unsupported GPU or DSL version; it never silently falls back to C++.
- Intermediate state checkpoints, zero-length packed sequences, non-BF16
  Q/K/V, head sizes other than 128, decode, and backward are intentionally out
  of scope and documented as unsupported.
- The changes are split by product responsibility: implementation,
  correctness/benchmark coverage, and documentation.

### CODE REVIEW

File: cula/ops/gdn/sm90/delta_rule.py [hunk 1]
```diff
@@ -0,0 +1,2399 @@
...
+        (1, num_v_heads * D, D),
+    )
+    o_tma = o.as_strided(
+        (D, total_seqlen, num_o_heads),
+        (1, num_o_heads * D, D),
+    )
+    total_checkpoints = state_checkpoints.shape[0] if needs_checkpointing else 1
+
+    workspace_size = get_device_sm_count(q.device) * 128
+    tensormaps_t = _get_cache_buf("gdn_prefill_tensormaps", workspace_size, q.device)
```
[the remark this answers is not in the document]
Comment (Aharrypotter) [reply cula/ops/gdn/sm90/delta_rule.py:2368]: Thanks for flagging this. I audited the exact specialization used by this PR and do not believe a per-launch workspace is required here.

The TensorMap workspace is indexed by the physical `smid()` with one 128-byte slot per SM. The source-bound H20 resource and occupancy capture for this 512-thread kernel reports 185,728 bytes of dynamic shared memory and exactly one active CTA per SM. In addition, the output S2G path executes `cp_async_bulk_wait_group(0)` before the CTA can exit.

Consequently, CTAs from different streams cannot be resident concurrently on the same SM and write the same slot. The next CTA can reuse that SM's slot only after the previous CTA's bulk operation has retired and the CTA has exited.

I added a comment in `6bc5c4d` to make this specialization-bound invariant explicit. If a future resource layout permits more than one resident CTA per SM, the cache design will need to be revisited.

File: cula/gdn/prefill.py [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+        cu_seqlens,
+        device=device,
+        ndim=1,
+        dtype=(torch.int32, torch.int64),
+    )
+    if cu_seqlens.numel() < 2:
+        raise ValueError("cu_seqlens must contain at least [0, total_tokens]")
+
+    def validate() -> tuple[int, ...]:
+        offsets = tuple(int(value) for value in cu_seqlens.detach().cpu().tolist())
```
[the remark this answers is not in the document]
Comment (Aharrypotter) [reply cula/gdn/prefill.py]: Fixed in `6bc5c4d`.

The default path no longer calls `cu_seqlens.detach().cpu().tolist()`. It derives `num_seqs` from tensor metadata (`numel() - 1`) and retains the existing rank, dtype, device, contiguity, and alignment checks.

CUDA-resident value checks are now an explicit `validate_inputs=True` diagnostic path, so production calls do not force a device-to-host synchronization.

An H20 CPU-profiler trace covered five calls with a new set of Q/K/V/g/beta/cu_seqlens/output tensors on every call and observed none of the following operations:

- `aten::_local_scalar_dense`
- `aten::_to_copy`
- `aten::all`
- `aten::gt`
- `aten::isfinite`
- `aten::item`

The complete SM90 test file also passes: 25 tests on H20 with CuTe DSL 4.5.1.

File: cula/gdn/prefill.py [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+    else:
+        _check_tensor("g", alpha, device=device, ndim=2, dtype=torch.float32)
+        if tuple(alpha.shape) != gate_shape:
+            raise ValueError(f"g must have shape {gate_shape}, got {tuple(alpha.shape)}")
+
+        def validate_alpha() -> None:
+            if not bool(torch.isfinite(alpha).all()) or not bool((alpha > 0).all()):
+                raise ValueError("g must contain finite, strictly positive forget factors")
+
+        _cached_content_validation("alpha", alpha, validate_alpha)
```
[the remark this answers is not in the document]
Comment (Aharrypotter) [reply cula/gdn/prefill.py]: I removed the default `bool(torch.isfinite(...).all())` and `bool((g > 0).all())` reductions, along with the object-identity content cache.

The default fast path now treats finite and positive CUDA values as caller preconditions, while `validate_inputs=True` retains the synchronous checks for debugging and input-pipeline validation. The API documentation and regression tests describe this contract.

The fresh-object Hopper profiler gate observed no scalar extraction or validation-reduction operations across five calls, and the complete SM90 test file passes 25/25.

File: cula/gdn/prefill.py [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+def is_sm90_gdn_prefill_available(
+    device: torch.device | int | str | None = None,
+) -> bool:
+    """Return whether the device and pinned CuTe DSL runtime can run GDN."""
+
+    if not torch.cuda.is_available():
+        return False
+    if device is None:
+        device = torch.cuda.current_device()
+    props = torch.cuda.get_device_properties(device)
```
[the remark this answers is not in the document]
Comment (Aharrypotter) [reply cula/gdn/prefill.py:98]: fixed

**Comment by Aharrypotter** [comment]
@icavan @zheyang0825 cc
